_posts/2024-02-19-jco-1.0.md

```diff
@@ -0,0 +1,270 @@
+
+```wit
+package my-org:wit-playground
+
+world playground {
+  export scream: func(input: string) -> string;
+}
+```
+
+We can save this to the generated `wit/world.wit` file [^cc-gen], and now we
+have defined a function `scream` which takes a string, and returns a string. And
+with that, it's time to write our Rust code!
+
+[^cc-gen]: If you're from the future: `cargo-component` generated this file for
+us in this location. Since it is not yet advertised as "stable", the output
+location for WIT files may change. If no `wit/world.wit` file was generated for
+you, please check a different location to see whether it might be somewhere else.
+
+### Writing a Rust library
+
+As mentioned, we're using the `cargo-component` tool here in this example to
+compile our Rust code to Wasm Components. It isn't yet at 1.0, but it *does*
+work pretty well. So we'll show you how you can use it today. If you're from the
+future (e.g. not February 2024), please check the `cargo-component` docs to see
+how to follow these steps. Roughly speaking what we're going to do is: take
+`.wit` definitions in, produce Rust code out. And then fill out the
+implementations. Even in the future there should be a way to do that.
+
+First things first: we want `cargo component` to generate the bindings for our
+WIT file. This is done via Rust traits, and it will do this automatically when
+we compile. This will yield a compiler error, but that's okay - we'll update the
+code in a second:
+
+```bash
+$ cargo component check
+```
+
+This will have generated Rust bindings, which on the current version of
+`cargo-component` is places under `lib/bindings.rs`. Don't worry about reading
+that code right now - it does a lot of heavy lifting so we don't have to. The
+important part is that it provides us with traits which we can implement. So
+let's implement those in our code, and write a little upper case function
+
+```rust
+mod bindings;                                     // Teach our program about `src/bindings.rs`
+
+use bindings::Guest;                              // Import the trait repr of our WIT world
+use wit_bindgen::rt::string::String as WitString; // Import the WIT string type
+
+struct Component;                                 // Define the type to export
+impl Guest for Component {                        // Import the WIT world + methods
+    fn scream(input: WitString) -> WitString {    // Implement our `scream` method
+        let mut s = input.to_uppercase();
+        s.push_str("!!1!");
+        s.into()
+    }
+}
+```
+
+Now if we re-run `cargo component check`, it should compile without any
+warnings. That means we're ready to run:
+
+```bash
+$ cargo component build      # Compile our Rust library to a Wasm Component
+```
+
+This will have written the generated Wasm code to
+`target/wasm32-wasi/debug/playground_jco.wasm`. Once Rust lands support for a
+native Preview 2 target, expect this to change to
```

Comment:
"native WASI 0.2.0 target"? dunno; but if we're using WASI 0.2.0 up top, keep it consistent to teach the readers the new language. One or the other.

You can also introduce the concept that "WASI Preview 2" is the same thing as "WASI 0.2.0" but we'll be using the latter going forward. Up to you.